[docs][pigment] Add example and guide section (#41249)
mui/material-ui: packages/zero-runtime/README.md

  - [Color schemes](#color-schemes)
  - [Switching color schemes](#switching-color-schemes)
  - [TypeScript](#typescript)
- [How-to guides](#how-to-guides)

>   - [Coming from Emotion or styled-components](#coming-from-emotion-or-styled-components)

Reviewer: ```suggestion
  - [Migrating from Emotion or styled-components](#coming-from-emotion-or-styled-components)
```

Maybe that's more obvious?

Author: I think I'd stick to `Coming` than `Migrating`. This section is to show some differences in term of mental model with examples, not a docs on how to migrate from emotion to zero.

Author: I got it from https://styled-components.com/docs/basics#coming-from-css

Reviewer: Makes sense — all good.

> ## Starter template

Reviewer: ```suggestion
## Starter templates
```

As we'll offer a Next.js _and_ Vite templates

Reviewer: Also, maybe this Starter templates section should be within the "Getting started"?

> Use the following commands to create a new project with the latest version of the framework and zero-runtime (for existing projects, see [Getting started](#getting-started)):

Author: ```suggestion
Use the following commands to create a new project with the latest framework version and zero-runtime (for existing projects, see [Getting started](#getting-started)):
```

});
```

> Note that the `props` function will not work if it is inside another closure, for example, inside an array.map:

Reviewer: ```suggestion
Note that the `props` function will not work if it is inside another closure, for example, inside an `array.map`:
```

  }
}
```

## How-to guides

### Coming from Emotion or styled-components

Emotion and styled-components are runtime CSS-in-JS libraries. What you write in your styles is what you get in the final bundle which means that the styles can be as dynamic as you want with the trade-offs of bundle size and performance overhead.

> Zero-runtime, on the other hand, extracts CSS at build time and replaces the JS code with hashed classnames and some CSS variables. This means that it has to know all of the styles to be extracted ahead of time, so there are rules and limitations that you need to be aware of when using javascript callbacks or variables in zero-runtime APIs.

Reviewer: ```suggestion
On the other hand, zero-runtime extracts CSS at build time and replaces the JS code with hashed class names and some CSS variables. This means that it has to know all of the styles to be extracted ahead of time, so there are rules and limitations that you need to be aware of when using JavaScript callbacks or variables in zero-runtime APIs.
```

Here are some common patterns and how to achieve them with zero-runtime:

1. A fixed set of styles

In Emotion or styled-components, you can use props to create styles conditionally:

```js
const Flex = styled('div')((props) => ({
  display: 'flex',
  ...(props.vertical // ❌ Zero Runtime will throw an error
    ? {
        flexDirection: 'column',
        paddingBlock: '1rem',
      }
    : {
        paddingInline: '1rem',
      }),
}));
```

But in zero-runtime, you need to define all of the styles ahead of time using the `variants` key:

```js
const Flex = styled('div')((props) => ({
  display: 'flex',
  variants: [
    {
      props: { vertical: true },
      style: {
        flexDirection: 'column',
        paddingBlock: '1rem',
      },
    },
    {
      props: { vertical: false },
      style: {
        paddingInline: '1rem',
      },
    },
  ],
}));
```

> > 💡 Keep in mind that `variants` is for fixed values of props, for example component's colors, sizes, states.

Reviewer: ```suggestion
> 💡 Keep in mind that the `variants` key is for fixed values of props, for example, a component's colors, sizes, and states.
```

> 2. Programatically generated styles

Author: ```suggestion
2. **Programatically generated styles**
```

> For emotion and styled-components, the styles will be different on each render and instances because the styles are generated at runtime:

Reviewer: ```suggestion
For Emotion and styled-components, the styles will be different on each render and instance because the styles are generated at runtime:
```

```js
function randomBetween(min: number, max: number) {
  return Math.floor(Math.random() * (max - min + 1) + min);
}
function generateBubbleVars() {
  return `
    --x: ${randomBetween(10, 90)}%;
    --y: ${randomBetween(15, 85)}%;
  `;
}

function App() {
  return (
    <div>
      {[...Array(10)].map((_, index) => (
        <div key={index} className={css`${generateBubbleVars()}`} />
      ))}
    </div>
  )
}
```

> However, in zero-runtime, all of the instances will have the same styles and won't change between renders because the styles are extracted at build time.

Reviewer: ```suggestion
However, in zero-runtime, all instances will have the same styles and won't change between renders because the styles are extracted at build time.
```
